[Woo POS] Coupons: Disallow adding duplicate coupons to cart

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

## Description

Partially addresses WOOMOB-259 by disallowing duplicated coupons to be added to the cart, while we get feedback from design on how to handle it UI-wise.

We also avoid calling the `pos_coupon_added_to_cart` track event multiple times by creating a new implementation for `POSItemActionHandler`, otherwise we could see the track event called multiple times despite the action being disallowed, as we used the same action handler across products and coupons:

```
🔵 Tracked pos_coupon_added_to_cart, properties: [blog_id: -1, is_wpcom_store: false, plan: , was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f]
🔵 Tracked pos_coupon_added_to_cart, properties: [is_wpcom_store: false, plan: , store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, was_ecommerce_trial: false, blog_id: -1, site_url: https://indiemelon.mystagingwebsite.com]
🔵 Tracked pos_coupon_added_to_cart, properties: [site_url: https://indiemelon.mystagingwebsite.com, store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, plan: , is_wpcom_store: false, blog_id: -1, was_ecommerce_trial: false]
```

## Testing information
- In POS, add a coupon to the cart
- Try again, observe how nothing happens. Removing it allows to be re-added.
- Confirm the `pos_coupon_added_to_cart` event is only logged once, when we add the coupon to the cart
- Confirm that adding products and variations to cart still work as expected.

## Screenshots

https://github.com/user-attachments/assets/6ee2133d-ddec-46ef-95d8-ca5c7aee2cfe

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Return early if same coupon ID is already in cart

iamgabrielma committed:
extract sut to be reusable within cart struct

We cannot extract the function to make SUT for the whole test suite (at least not straight-away) as crashes due analytics running on different threads that we expect for each enum in the test suite.

iamgabrielma committed:
Disallow tracking add to cart event when is disallowed

Despite disallowing for duplicated coupons to be added to cart, the action handler is generic to all items so we were still calling the tracking event. We create a handler specifically for coupons to avoid calling the event when it shouldnt

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
lint

staskus commented:
@iamgabrielma thanks for the work!

Let's wait before merging until we get input from design. I think it's better right now to show an error in the checkout rather than not allow tapping the item without any visual signal.

Also, let's clarify if the item should be tappable when it's disabled (including the expired one). Maybe we wouldn't need a separate item handler logic.

Thanks!

iamgabrielma commented:
Confirmed with design FnDpCHVQ3Zpefad04ptbd4-fi-node-id=419-79264#1230020091 that nothing should happen when attempting to add a duplicated coupon to the cart. Will resolve the merge conflicts EOD and re-submit 👍

iamgabrielma committed:
Merge branch 'trunk' into task/woomob-259-disallow-adding-same-coupon-twice

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/ItemListView.swift
#	WooCommerce/WooCommerceTests/POS/Models/PointOfSaleAggregateModelTests.swift

iamgabrielma committed:
Delete unused makeSUT

Not that reusable, as analytics is not thread safe and crashes randomly some tests. For now let’s declare the dependencies per test basis

iamgabrielma committed:
restore using analytics shared dependency for CartTests

staskus commented:
@iamgabrielma to confirm, is it ready for review?

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/POSItemActionHandler.swift:
```diff
@@ -51,6 +51,28 @@ final class StandardPOSItemActionHandler: POSItemActionHandler {
 }
 
+@available(iOS 17.0, *)
+final class CouponsItemActionHandler: POSItemActionHandler {
```
No longer needed, can be removed 👍

iamgabrielma replied:
Thanks, I didn't see that we stopped using `CouponsItemActionHandler` since I worked on this bit. Removing it introduces a regression thought, as we treat products and coupons equally when dealing with track events, meaning: We will track the `add_to_cart` event despite coupons not being added to cart when are duplicated:

```
🔵 Tracked pos_coupon_added_to_cart, properties: [site_url: https://indiemelon.mystagingwebsite.com, blog_id: -1, was_ecommerce_trial: false, plan: , store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, is_wpcom_store: false]
🔵 Tracked pos_coupon_added_to_cart, properties: [store_id: c5bd46cc-1804-4f7b-badb-bb98c449127f, plan: , is_wpcom_store: false, blog_id: -1, was_ecommerce_trial: false, site_url: https://indiemelon.mystagingwebsite.com]
```

I've added a small [helper function](https://github.com/woocommerce/woocommerce-ios/pull/15551/commits/3bfe659f289f4e5e3453c98580968677510466c1) to handle this case in `POSItemActionHandler`, as this will happen both when we add coupons via list and via search.

I'll leave this one without merging for now in case you want to take a look, in a nutshell when adding coupons from either list or search, the track event should be logged just once.

staskus reviewed (approved):
Thanks! Just remove the unused code 👍

iamgabrielma commented:
> to confirm, is it ready for review?

Yes! Sorry I wasn't clear 🙇

iamgabrielma committed:
Remove unused CouponsItemActionHandler

Since the specific CouponsItemActionHandler is not used anymore, we still need to deal with duplicates coupons in the general handler to avoid duplicated track events

staskus commented on WooCommerce/Classes/POS/Models/Cart.swift:
```diff
@@ -49,7 +49,9 @@ extension Cart {
             return
         case .coupon(let coupon):
-            let couponItem = Cart.CouponItem(id: UUID(), code: coupon.code, summary: coupon.summary)
+            guard !coupons.contains(where: { $0.id == coupon.id }) else { return }
```
Do we need this check here if we skip the duplicate in the action handler?

iamgabrielma replied:
We don't anymore, I've removed this on https://github.com/woocommerce/woocommerce-ios/pull/15551/commits/68096bc0ddf9c9ddc9efa60df3e378fe25f89757 and added initial tests for the action handler on https://github.com/woocommerce/woocommerce-ios/pull/15551/commits/a7c59776e0c0033bfb34bdb4f9be5f1b9dae34e1

iamgabrielma committed:
remove early return duplication

staskus reviewed (approved):
👍

iamgabrielma committed:
Add POSItemActionHandlerTests

iamgabrielma committed:
extract MockPOSSearchHistoryService

iamgabrielma committed:
use specific handler type when testing search
